spacy/util.py

```diff
@@ -256,20 +258,85 @@ def lang_class_is_loaded(lang: str) -> bool:
     return lang in registry.languages
 
 
+def find_matching_language(lang: str) -> Optional[str]:
+    """
+    Given an IETF language code, find a supported spaCy language that is a
+    close match for it (according to Unicode CLDR language-matching rules).
+    This allows for language aliases, ISO 639-2 codes, more detailed language
+    tags, and close matches.
+
+    Returns the language code if a matching language is available, or None
+    if there is no matching language.
+
+    >>> find_matching_language('en')
+    'en'
+    >>> find_matching_language('pt-BR')  # Brazilian Portuguese
+    'pt'
+    >>> find_matching_language('fra')  # an ISO 639-2 code for French
+    'fr'
+    >>> find_matching_language('iw')  # obsolete alias for Hebrew
+    'he'
+    >>> find_matching_language('no')  # Norwegian
+    'nb'
+    >>> find_matching_language('mo')  # old code for ro-MD
+    'ro'
+    >>> find_matching_language('zh-Hans')  # Simplified Chinese
+    'zh'
+    >>> find_matching_language('zxx')
+    None
+    """
+    import spacy.lang  # noqa: F401
+    if lang == 'xx':
+        return 'xx'
+
+    # Find out which language modules we have
+    possible_languages = []
+    for modinfo in pkgutil.iter_modules(spacy.lang.__path__):
+        code = modinfo.name
+        if code == 'xx':
+            # Temporarily make 'xx' into a valid language code
+            code = 'mul'
+        try:
+            language = langcodes.Language.get(code)
+            if language.is_valid():
+                possible_languages.append(code)
+        except langcodes.tag_parser.LanguageTagError:
+            pass
+
+    match, _dist = langcodes.closest_match(
+        lang, possible_languages, max_distance=10
+    )
```

Comment:
Will the 10 here always be fine, or could it come back to haunt us later?

Reply:
I've used a distance of 10 as my threshold for matching for a while, but CLDR changes its language distances a lot, so I can't promise anything.

The most distant match this currently allows is from a deprecated code for Montenegrin (`cnr`) to Croatian (`hr`), a distance of 9. I currently want to rely on matching Bosnian to Croatian, which is a distance of 8.

I can see a vague reason to be worried about a distance of 10: in much earlier versions of CLDR, they would give a distance of 10 from a language to the world language it's being colonized by, like from Wolof to French or from Tamil to English. That was clearly unexpected and unreasonable. They have greatly increased those language distances now (to 34 and 44 respectively). I don't think they'll roll back to being unreasonable, but I could change the maximum to 9 just to be sure?

Reply:
I iterated through CLDR's supported language tags and found one at a distance of 10: `yue-Hans` (Cantonese in simplified characters, as used in Guangzhou) matches `zh` at distance 10. So that's the most borderline case.

Reply:
also, I control updates to langcodes, so if I think that CLDR's language distances have drastically changed in meaning I could save that update for a major version.

Reply:
It looks like the default in `langcodes` is 25, that seems pretty high after reading your explanation here? I'd personally prefer not specifying this parameter in `spacy` and just using the default, but I can see that too high a value would be problematic. If it won't be updated in `langcodes`, I think having a more restrictive setting in spaCy would be good, especially when publishing this feature as users will not expect some kind of fuzzy matching behaviour here, and a typo might come to bite them.

Reply:
That actually brings up the question whether there would be users who'd like to be able to enforce the current, strict behaviour. If we want to cater for that use-case, I think having this variable be adjustable (env var?) would work.

Reply:
I think we'd just be setting up for confusion by making that configurable. For what it's worth, here is the complete list of plain language codes that are within a distance of 10 of one of our languages, but not 0:

`['bs', 'gsw', 'no']`

There are a lot more cases that involve regional codes like `en-UK` but I don't think those would come from a typo.

25 is a high default, yeah. I picked 25 as a default for the simple case where i18n consists of looking up strings in a table (which is what langcodes uses for naming languages when you include the optional `language_data` package). There are some distances near 25 that we definitely wouldn't want for NLP.